### ISSUE TITLE
Serialization of SubArrays: reconsider trimming

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#14708]
In #14625, there was some unresolved discussion of how (de)serialization of SubArrays should work.

There are two "easy" choices:

- Our current practice is equivalent to decreeing that the SubArray is equivalent (for the purposes of serialization) to its in-bounds contents---serialization "trims" any data in the parent falling outside the in-bounds region. This is efficient, but would fail for code that expects to access out-of-bounds regions of the parent array.
- Another option would decree that a SubArray should preserve every detail of the original view. This would be equivalent to deleting [this method](https://github.com/JuliaLang/julia/blob/8801efadadf0e1301ee6449ebd31bcba6f3df865/base/serialize.jl#L194-L197), and having the entire parent be serialized whenever a view of that parent is serialized. This would allow code to access out-of-bounds regions of the parent, at the cost of potentially serializing much more data than will be used. However, serializing both the parent and its view, or serializing two SubArrays sharing the same parent, would not preserve any of the relationships between them, and would likely result in double-serialization (and deserialization) of the raw data.

While it would be more difficult, it's also worth considering whether it's possible to preserve relationships. This does not seem entirely out of reach, thanks to the use of [`SerializationState`](https://github.com/JuliaLang/julia/blob/8801efadadf0e1301ee6449ebd31bcba6f3df865/base/dict.jl#L379-L384), which sets up an `ObjectIdDict`. However, this object is not typically expected to be persistent across separate calls to `serialize`, and for calls like `remotecall_fetch` does not even appear explicitly. Thus, implementation also appears to involve API questions.

**Comment by eschnett** [comment]
What is the reason for having a SubArray with array bounds if it can be accessed outside these bounds? Is this about maintaining backward compatibility -- some algorithms use subarrays, but since their bounds are not checked, the subarray's bounds are not declared correctly? Or is this about cases where Julia's array bounds declarations are insufficient -- do people e.g. want to access `a[-2]`, which happens to work fine with subarrays, but one cannot declare bounds that would permit this?

If I have an array that is not going to be modified, then I'd use subarrays as a more efficient version of making a copy of part of an array. When I pass this subarray to another function, I expect this function to only access the part I passed. I would thus view accessing a subarray out of bounds as error.

Passing a small portion of an array to another process is a very common operation, e.g. for a ghost-zone exchange, or when dividing an array onto workers. It would be impractical if Julia's default was to transmit the whole array.

**Comment by nalimilan** [comment]
OTOH, preserving the coupling between the parent and its views would make the semantics more predictable, as @simonster noted on the other issue. It sounds like a good idea that moving objects to another worker doesn't change the behavior of code that would mutate the view and expect it to affect the parent.

Also, it's always easy to call `copy` if you don't care about the reference to the parent array, while the contrary isn't true.

**Comment by timholy** [comment]
@eschnett, as you note, right now we don't have bounds-checking on accesses to data from SubArrays (although thanks to @blakejohnson now we can implement it). There may be code out there that depends on current behavior. Moreover, any code that exploits the internal representation of a SubArray also might be interested in "padding," for example (I've made use of that myself in several places).

But personally, I agree that the default for SubArrays should probably be just about the in-bounds portion, and that the performance hit from serializing the parent could be astronomical.

@nalimilan,
> OTOH, preserving the coupling between the parent and its views would make the semantics more predictable, as @simonster noted on the other issue. It sounds like a good idea that moving objects to another worker doesn't change the behavior of code that would mutate the view and expect it to affect the parent.

Just to be clear, currently I have no good ideas for how one might achieve that, so this is a pretty speculative "option."

> Also, it's always easy to call copy if you don't care about the reference to the parent array, while the contrary isn't true.

That is true, and an important point. Worth noting also that when you have SubArrays fields of a container, it gets a little more complicated.

**Comment by nalimilan** [comment]
Of course, I wasn't considering the difficulties in implementing this -- merely looking at what would be desirable. ;-)

**Comment by eschnett** [comment]
Desirable:
- if both the subarray and the original array are serialized, the relationship should be preserved
- if only the subarray is serialized, and there is no other overlapping serialized subarray, then the original array does not need to be serialized

Regarding "easily calling copy": I'd like to avoid copying a subarray just to improve serialization. If push comes to shove, I'd probably implement my own serialization instead.
